File: machine_maintenance/machine_maintenance/report/machine_maintenance_report/machine_maintenance_report.py
# Copyright (c) 2025, karthik polisetty and contributors
# For license information, please see license.txt


import frappe
from frappe.utils import flt

def execute(filters=None):
    filters = filters or {}

    consolidated = bool(filters.get("consolidated"))
    columns = get_columns(consolidated)
    data = get_data(filters, consolidated)

    return columns, data


# -------------------------------------------------------
# Columns
# -------------------------------------------------------
def get_columns(consolidated):
    cols = [
        {
            "label": "Machine",
            "fieldname": "machine_name",
            "fieldtype": "Link",
            "options": "Item",
            "width": 200,
        }
    ]

    if not consolidated:
        cols += [
            {
                "label": "Maintenance Date",
                "fieldname": "maintenance_date",
                "fieldtype": "Date",
                "width": 110,
            },
            {
                "label": "Technician",
                "fieldname": "technician",
                "fieldtype": "Link",
                "options": "Employee",
                "width": 150,
            },
            {
                "label": "Status",
                "fieldname": "status",
                "fieldtype": "Data",
                "width": 110,
            },
        ]
    else:
        # Only show grouped status column
        cols.append({
            "label": "Status",
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 110,
        })

    # Common column
    cols.append({
        "label": "Total Cost",
        "fieldname": "total_cost",
        "fieldtype": "Currency",
        "width": 140,
    })

    return cols


# -------------------------------------------------------
# SQL Filters
# -------------------------------------------------------
def get_conditions(filters):
    conditions = ["docstatus = 1"]
    values = {}

    if filters.get("machine"):
        conditions.append("machine_name = %(machine)s")
        values["machine"] = filters["machine"]

    if filters.get("technician"):
        conditions.append("technician = %(technician)s")
        values["technician"] = filters["technician"]

    if filters.get("from_date"):
        conditions.append("maintenance_date >= %(from_date)s")
        values["from_date"] = filters["from_date"]

    if filters.get("to_date"):
        conditions.append("maintenance_date <= %(to_date)s")
        values["to_date"] = filters["to_date"]

    return " AND ".join(conditions), values


# -------------------------------------------------------
# Fetch Data
# -------------------------------------------------------
def get_data(filters, consolidated):
    conditions, values = get_conditions(filters)

    if consolidated:
        query = f"""
            SELECT
                machine_name,
                'Consolidated' AS status,
                SUM(cost) AS total_cost
            FROM `tabMachine Maintenance`
            WHERE {conditions}
            GROUP BY machine_name
            ORDER BY machine_name
        """
    else:
        query = f"""
            SELECT
                machine_name,
                maintenance_date,
                technician,
                status,
                cost AS total_cost
            FROM `tabMachine Maintenance`
            WHERE {conditions}
            ORDER BY maintenance_date DESC, machine_name
        """

    rows = frappe.db.sql(query, values, as_dict=True)

    # Ensure cost always numeric
    for row in rows:
        row["total_cost"] = flt(row.get("total_cost"))

    return rows

File: machine_maintenance/machine_maitenance/report/machine_maintenance_report/machine_maintenance_report.js
// Copyright (c) 2025, karthik polisetty and contributors
// For license information, please see license.txt

// Copyright (c) 2025, karthik polisetty and contributors
// For license information, please see license.txt

frappe.query_reports["Machine Maintenance Report"] = {
    "filters": [
        {
            "fieldname": "machine",
            "label": __("Machine"),
            "fieldtype": "Link",
            "options": "Item"
        },
        {
            "fieldname": "technician",
            "label": __("Technician"),
            "fieldtype": "Link",
            "options": "Employee"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date"
        },
        {
            "fieldname": "consolidated",
            "label": __("Consolidated"),
            "fieldtype": "Check",
            "default": 0
        }
    ],

    formatter: function (value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (!data) return value;
        if (data.status === "Consolidated") {
            return value;
        }

        let status = data.status;

        if (status === "Overdue") {
            return `<span style="background-color:#ffcccc; display:block;">${value}</span>`;
        }
        if (status === "Scheduled") {
            return `<span style="background-color:#fff3cd; display:block;">${value}</span>`;
        }
        if (status === "Completed") {
            return `<span style="background-color:#d4edda; display:block;">${value}</span>`;
        }

        return value;
    }
};
